tests/structs/test_query_output_details.py

```diff
@@ -15,37 +15,47 @@ def setUp(self) -> None:
-            ("no properties given, output type given", False, "TO_OBJECT_LIST"),
-            ("properties and output type given", True, "TO_LIST"),
+            ("none given", {}),
+            ("props given", {"properties_to_select": ["server_id", "server_name"]}),
+            ("output type given", {"output_type": "TO_LIST"}),
+            ("group by given", {"group_by": "server_name"}),
+            ("sort by given", {"sort_by": ["server_name", "server_id"]}),
         ]
     )
-    def test_from_kwargs(self, _, set_mock_props, mock_output_type):
+    def test_from_kwargs(self, _, mock_kwargs):
         """
         tests that from_kwargs static method works expectedly
-        should iteratively convert properties into given enums class and output_type into output type enum.
-        set correct attributes in QueryOutputDetails dataclass and return
+        method should create a QueryOutputDetails with default params
         """
-
-        expected_mock_props = list(ServerProperties)
-        mock_props = [prop.name for prop in ServerProperties]
-        if set_mock_props:
-            expected_mock_props = [
-                ServerProperties.SERVER_ID,
-                ServerProperties.SERVER_NAME,
-            ]
-            mock_props = ["server_id", "server_name"]
-
-        expected_mock_output_type = QueryOutputTypes.TO_STR
-        if mock_output_type:
-            expected_mock_output_type = QueryOutputTypes[mock_output_type]
-
-        res = QueryOutputDetails.from_kwargs(
-            prop_cls=ServerProperties,
-            properties_to_select=mock_props,
-            output_type=mock_output_type,
-        )
-
-        assert set(res.properties_to_select) == set(expected_mock_props)
-        assert res.output_type == expected_mock_output_type
+        out = self.instance.from_kwargs(prop_cls=ServerProperties, **mock_kwargs)
+        if "properties_to_select" not in mock_kwargs.keys():
+            self.assertEqual(out.properties_to_select, list(ServerProperties))
+        else:
+            self.assertEqual(
+                out.properties_to_select,
+                [ServerProperties.SERVER_ID, ServerProperties.SERVER_NAME],
+            )
+
+        if "output_type" not in mock_kwargs.keys():
+            self.assertEqual(out.output_type, QueryOutputTypes.TO_STR)
+        else:
+            self.assertEqual(out.output_type, QueryOutputTypes.TO_LIST)
+
+        if "group_by" not in mock_kwargs.keys():
+            self.assertEqual(out.group_by, None)
+        else:
+            self.assertEqual(out.group_by, ServerProperties.SERVER_NAME)
+
+        if "sort_by" not in mock_kwargs.keys():
+            self.assertEqual(out.sort_by, None)
+        else:
+            self.assertEqual(
+                out.sort_by,
+                [
+                    (ServerProperties.SERVER_NAME, False),
+                    (ServerProperties.SERVER_ID, False),
+                ],
+            )
```

Comment:
Ditto about this being cases, building on above just return and compare in the test case for only the thing your testing. I.e. don't check `group_by` in the `sort_by` cases

```
def test_from_kwargs_sort_given(self):
   out = self._run_from_kwargs_case({"sort_by": ["server_name", "server_id"]})
   
  self.assertEqual(out.sort_by, [ ... ])

def test_from_kwargs_sort_missing(self):
   out = self._run_from_kwargs_case({})
   
  self.assertEqual(out.sort_by, None)
```

If changing one implies another (e.g. SortBy changes GroupBy) add a third class of test cases to check this but it makes testing with state machines much easier

```
def test_setting_sort_by_resets_group_by
```